odoo/api.py

```diff
@@ -726,19 +726,20 @@ def reset(cls):
         """
         cls._local.environments = Environments()
 
-    def __new__(cls, cr, uid, context):
+    def __new__(cls, cr, uid, context, real_uid=None):
         assert context is not None
         args = (cr, uid, context)
 
         # if env already exists, return it
         env, envs = None, cls.envs
         for env in envs:
-            if env.args == args:
+            if env.args == args and env.real_uid == real_uid:
```

Comment:
`env.real_uid` is always an integer. The default value of real_uid is None, which will never match an existing environment.
The calculation of real uid on line 742 should be moved up to line 732:

`real_uid = uid if real_uid is None else real_uid`

(Attribute assignment into self should remain at 742.)